docs/design/ha_master.md

```diff
@@ -0,0 +1,188 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<!-- TAG RELEASE_LINK, added by the munger automatically -->
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.3/docs/design/access.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Automated HA master deployment (in GCE)
+
+**Author:** filipg@, jsz@
+
+# Introduction
+We want to allow users to easily replicate kubernetes masters to have highly available cluster,
+using automated tools like kube-up.sh and kube-down.sh.
+
+This document describes technical design of this feature. It assumes that we are using aforementioned
+scripts for cluster deployment. It focuses on GCE use-case but all of the ideas described in
+the following sections should be easy to port to AWS, other cloud providers and bare-metal environment.
+
+# Overview
+In a cluster with replicated master machine we will have the following component interactions:
+* All etcd replicas will be clustered together and will be using master election
+and quorum mechanism to agree on the state. All of these mechanisms are integral
+parts of etcd and we will only have to configure them properly.
+* All apiserver replicas will be working independently talking to a etcd on the
+same machine, which if needed will forward requests to the current etcd master
+(as explained (here)[https://coreos.com/etcd/docs/latest/getting-started-with-etcd.html]);
+in front of apiservers we will have an L4 load balancer to spread the load among all the replicas.
+* Controller manager, scheduler & cluster autoscaler will use lease mechanism and
+only a single instance will be an active master. All other will be waiting in a standby mode.
+* All add-on managers will work independently and each of them will try to keep add-ons in sync
+
+# Detailed design
+
+## VM naming
+VMs with master replicas will have the following names:
+* first VM - `${MASTER_NAME}`
+* next VMs - `${MASTER_NAME}-<hash>`, where `<hash>` is a 3-characters long alphanumeric string.
```

Comment:
I see your point. This section is actually not required, because I'm not using those names anywhere. Everywhere else I'm using only IP addresses.

I can rename this section to clearly say it's about GCE.